packages/components/src/custom-select-control-v2/legacy-component/index.tsx

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import * as Ariakit from '@ariakit/react';
+/**
+ * Internal dependencies
+ */
+import _CustomSelect from '../custom-select';
+import type { LegacyCustomSelectProps } from '../types';
+import { CustomSelectItem } from '..';
+
+function _LegacyCustomSelect( props: LegacyCustomSelectProps ) {
+	// Forward props + store from v2 implementation
+	const store = Ariakit.useSelectStore( {
+		async setValue( value ) {
+			if ( ! props.onChange ) return;
+
+			// Executes the logic in a microtask after the popup is closed.
+			// This is simply to ensure the isOpen state matches that in Downshift.
+			await Promise.resolve();
+			const state = store.getState();
+
+			const changeObject = {
+				selectedItem: {
+					// value will always be a string for the legacy component
+					name: value as string,
+					key: state.activeId as string,
+				},
+				highlightedIndex: state.renderedItems.findIndex(
+					( item ) => item.value === value
+				),
+				isOpen: state.open,
+			};
+			props.onChange( changeObject );
```

Comment:
Just to note, I'm still looking into the onChange return object. I've implemented the suggestion by @diegohaz from an earlier PR: https://github.com/WordPress/gutenberg/pull/55234#discussion_r1361394388, minus the `type` and `inputValue`. I'm looking to see if it makes sense to include these / how we could implement them. 

The value of `type` is the interaction method used to make a selection, i.e. `__item_click__`, `__menu_keydown_enter__`

The value of `inputValue` is the first character entered via keydown to make a selection (with `type` returning `__togglebutton_keydown_character__`). If keydown isn't used to make a selection, then it returns an empty string.

Reply:
I'd keep the `type` as `__item_click__` by default for simplicity's sake. If anyone depends on these values, it's likely the type they're handling as standard.

You could get this value by listening to DOM events and storing the information in a ref for subsequent use in Ariakit's `setValue` prop. But honestly, I wouldn't sweat it. It's not even mentioned in the Downshift docs. It doesn't seem to be part of their public API.